tools/column_ordering/column_ordering.rb

```diff
@@ -0,0 +1,231 @@
+
+  def initialize(table, connection)
+    @table = table
+    @connection = connection
+    raise ColumnOrderingError, "The database does not contain table #{table}" unless connection.tables.include?(table)
+  end
+
+  def expected_columns
+    @expected_columns ||= YAML.load_file(SCHEMA_FILE)[table]
+  end
+
+  def current_columns
+    @connection.columns(table).map(&:name)
+  end
+
+  def ordering_okay?
+    expected_columns == current_columns
+  end
+
+  def fix_column_ordering
+    assert_column_list_sizes_match! && assert_column_list_contents_match!
+
+    if ordering_okay?
+      puts "The columns of #{table} are correctly ordered"
+      return
+    end
+
+    create_table_statement, remaining_schema_structure = parse_create_table
+
+    begin
+      @connection.begin_db_transaction
+
+      puts "Retrieving old sequence value ..."
+      pk, sequence = @connection.pk_and_sequence_for(table)
+      raise ColumnOrderingError, "Failed to retrieve primary key and sequence for #{table}" unless pk && sequence
+      sequence_value = @connection.select_value("SELECT nextval('#{sequence}')")
+      puts "Retrieved old sequence value #{sequence_value}"
+
+      puts "Renaming old table ..."
+      @connection.rename_table(table, "#{table}_old")
+      puts "Table renamed"
+
+      puts "Creating new table ..."
+      create_new_table = reordered_create_table_statement(create_table_statement)
+      @connection.exec_query(create_new_table)
+      puts "New table created as:\n#{create_new_table}"
+
+      puts "Moving data into new table ..."
+      @connection.exec_query(<<-SQL)
+        INSERT INTO #{table} SELECT #{expected_columns.join(",")} FROM #{table}_old
+      SQL
+      puts "Data moved"
+
+      puts "Dropping old table ..."
+      @connection.exec_query("DROP TABLE #{table}_old")
+      puts "Old table removed"
+
+      puts "Recreating remaining schema structure ..."
+      @connection.execute(remaining_schema_structure).check
+      puts "Schema structure recreated"
+
+      puts "Setting new sequence to old value ..."
+      @connection.set_pk_sequence!(table, sequence_value)
+      puts "Sequence value set"
+    rescue
+      @connection.exec_rollback_db_transaction
+      raise
+    end
+
+    @connection.commit_db_transaction
+    puts "Columns of table #{table} successfully reordered"
+  end
+
+  def table_dump
+    connection_params = @connection.raw_connection.conninfo_hash
+    conf = {
+      :dbname   => connection_params[:dbname],
+      :host     => connection_params[:host],
+      :user     => connection_params[:user],
+      :password => connection_params[:password],
+      :port     => connection_params[:port]
+    }.delete_blanks
+
+    params = {
+      :s => nil,
+      :t => table,
+      :d => PG::Connection.parse_connect_args(conf)
+    }
+
+    pg_dump_result = AwesomeSpawn.run("pg_dump", :params => params)
+
+    raise ColumnOrderingError <<-ERROR.gsub!(/^ +/, "") if pg_dump_result.failure?
```

Comment:
Would we rather have the heredoc string all left aligned? I think it's easier to read this way.

This will all go away with ruby 2.3 though :smile:  https://infinum.co/the-capsized-eight/articles/multiline-strings-ruby-2-3-0-the-squiggly-heredoc